frame/curve-amm/src/lib.rs

```diff
@@ -1011,53 +656,73 @@ pub mod pallet {
 			None
 		}
 
-		fn transfer_liquidity_into_pool(
-			pool_account_id: &T::AccountId,
-			pool: &mut PoolInfo<T::AccountId, T::AssetId, T::Balance>,
-			source: &T::AccountId,
-			destination_asset_index: usize,
-			amount: T::Balance,
-		) -> DispatchResult {
-			T::LpToken::transfer(
-				pool.assets[destination_asset_index],
-				source,
-				pool_account_id,
-				amount,
-				true,
-			)?;
-
-			pool.total_balances[destination_asset_index] = pool.total_balances
-				[destination_asset_index]
-				.checked_add(&amount)
-				.ok_or(Error::<T>::InconsistentStorage)?;
-
-			Ok(())
-		}
-
-		fn transfer_liquidity_from_pool(
-			pool_account_id: &T::AccountId,
-			pool: &mut PoolInfo<T::AccountId, T::AssetId, T::Balance>,
-			source_asset_index: usize,
-			destination: &T::AccountId,
-			amount: T::Balance,
-		) -> DispatchResult {
-			T::LpToken::transfer(
-				pool.assets[source_asset_index],
-				pool_account_id,
-				destination,
-				amount,
-				true,
-			)?;
-
-			pool.total_balances[source_asset_index] = pool.total_balances[source_asset_index]
-				.checked_sub(&amount)
-				.ok_or(Error::<T>::InconsistentStorage)?;
-
-			Ok(())
-		}
-
 		pub fn account_id(pool_id: &PoolId) -> T::AccountId {
 			PALLET_ID.into_sub_account(pool_id)
 		}
+
+		pub fn initialize_block(block_number: T::BlockNumber) {
+			with_transaction(|| {
+				let results = Pools::<T>::iter()
+					.map(|(pool_id, pool_info)| {
+						let pool_account_id = Self::account_id(&pool_id);
+						for vault_id in pool_info.assets_vault_ids {
+							let asset_id = <T::Vault>::asset_id(&vault_id)?;
+							match <T::Vault as StrategicVault>::available_funds(
+								&vault_id,
+								&pool_account_id,
+							)? {
+								FundsAvailability::Depositable(balance) => {
+									let balance = <T as Config>::LpToken::reducible_balance(
+										asset_id,
+										&pool_account_id,
+										false,
+									)
+									.min(balance);
+									<T::Vault as StrategicVault>::deposit(
+										&vault_id,
+										&pool_account_id,
+										balance,
+									)?
+								},
+								FundsAvailability::Withdrawable(balance) =>
+									<T::Vault as StrategicVault>::withdraw(
+										&vault_id,
+										&pool_account_id,
+										balance,
+									)?,
```

Comment:
Same as previous @dzmitry-lahoda comment, you cannot withdraw everything as you might unbalance yourself, instead, you should try to rebalance and if already balanced, take from each vault an amount which is equal to the minimum you can withdraw from one of the vaults. Also, can we extract this to its own function and add valid/invalid tests for them please?

Reply:
@hussein-aitlahcen  so in pool creation should I ask fund from vaults? or both values 0 means it is balanced ?